pyquil/quilatom.py

```diff
@@ -643,6 +643,12 @@ def __init__(self, name, params):
         self.name = name
         self.params = params
 
+    def __hash__(self):
```

Comment:
If you're providing your own `__hash__` method, you usually want to define `__eq__` as well. For example, from the python [docs](https://docs.python.org/3/reference/datamodel.html#object.__hash__)

> If a class does not define an `__eq__()` method it should not define a `__hash__()` operation either.

Also, unless there is some reason you want this specific hash function, another idiom is to just stuff everything in a tuple and hash that, i.e. `return hash((self.name, *self.params.items()))`. For example, `MemoryReference.__hash__`, above, is implemented this way. Python's [`tuplehash`](https://github.com/python/cpython/blob/master/Objects/tupleobject.c#L368) is usually faster and provides better collision resilience.

Nit: the docstring above claims that `params` is a `list` but looks like it's actually a `dict`.